PVYabause/yabause/src/gtk/doc/yabause.1
.TH YABAUSE 1 "November 22, 2014" "yabause-0.9.14"
.SH NAME
yabause \- Yet Another Buggy And Uncomplete Saturn Emulator
.SH SYNOPSIS
.B yabause
[ \fB\-afh\fP ] [ \fB\-ns\fP ] [ \fB\-b <BIOS>\fP ] [ \fB\-c <CDROM>\fP ] [ \fB\-i <ISO>\fP ]
[ \fB\-\-binary=<BINARY>[:<ADDRESS>]\fP ]
.SH DESCRIPTION
\fBYabause\fP is a Sega Saturn emulator. \fBYabause\fP needs either a bios file, a game or a binary to run.
Games can be loaded from a real cd device or from dump files.
.SH OPTIONS
.TP
.BI \-a
.TP
.BI \-\-autostart
Automatically start emulation.
.TP
.BI \-\-autoframeskip=0|1
Enable or disable auto frame skipping / limiting.
.TP
.BI \-\-autoload
Automatically start emulation and load a save state.
.TP
.BI \-b
.TP
.BI \-\-bios
Choose a bios file.
.TP
.BI \-\-binary
Use a binary file. Content of the file will be loaded to 0x06004000 and execution will start from that address.
You can provide an alternative address to load and run the binary with the \-\-binary=<BINARY>:<ADDRESS> syntax.
When using this option, emulation is automatically started and you shouldn't use it in cunjunction with \-a.
This option is intended for homebrew developers wanting to test their programs in \fBYabause\fP.
.TP
.BI \-c
.TP
.BI \-\-cdrom
Choose the cdrom device.
.TP
.BI \-f
.TP
.BI \-\-fullscreen
Start the emulator in fullscreen.
.TP
.BI \-h
Display a short help text.
.TP
.BI \-i
.TP
.BI \-\-iso
Choose a dump file. The dump can be either in iso or bin/cue file format.
.TP
.BI \-ns
.TP
.BI \-\-nosound
Turns sound off. This option actually set the sound core to the dummy one.
.SH FILES
.TP
\fB<BINARY>\fR
An executable for the Saturn, usually with a BIN extension.
.TP
\fB<BIOS>\fR
A Saturn ROM BIOS image.
.TP
\fB<CDROM>\fR
A CDROM device file.
.TP
\fB<ISO>\fR
A Saturn game dump.
.SH AUTHORS
Copyright (c) 2002-2010 Yabause Team

Web:	http://yabause.org

Please don't ask for roms, bios files or any other
copyrighted stuff.
.SH COPYRIGHT
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as 
published by the Free Software Foundation; either version 2 of 
the License, or (at your option) any later version.

This program is distributed in the hope that it will be
useful,but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public
License along with this program; if not, write to the Free
Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA 02110-1301  USA

See the GNU General Public License details in COPYING.


PVYabause/yabause/src/android/src/org/yabause/android/Yabause.java
/*  Copyright 2011-2013 Guillaume Duhamel

    This file is part of Yabause.

    Yabause is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Yabause is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Yabause; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
*/

package org.yabause.android;

import java.lang.Runnable;
import java.io.File;
import java.io.FileOutputStream;

import android.app.Activity;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MenuInflater;
import android.view.KeyEvent;
import android.app.Dialog;
import android.app.AlertDialog;
import android.app.AlertDialog.Builder;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.ContentValues;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.view.View;
import android.graphics.Bitmap;
import android.os.Environment;
import android.provider.MediaStore;
import android.net.Uri;
import org.yabause.android.GameInfo;

class InputHandler extends Handler {
    private YabauseRunnable yr;

    public InputHandler(YabauseRunnable yr) {
        this.yr = yr;
    }

    public void handleMessage(Message msg) {
        Log.v("Yabause", "received message: " + msg.arg1 + " " + msg.arg2);
        if (msg.arg1 == 0) {
            yr.press(msg.arg2);
        } else if (msg.arg1 == 1) {
            yr.release(msg.arg2);
        }
    }
}

class YabauseRunnable implements Runnable
{
    public static native int init(Yabause yabause);
    public static native void deinit();
    public static native void exec();
    public static native void press(int key);
    public static native void release(int key);
    public static native int initViewport();
    public static native int cleanup();
    public static native int drawScreen();
    public static native int lockGL();
    public static native int unlockGL();
    public static native void enableFPS(int enable);
    public static native void enableFrameskip(int enable);
    public static native void setVolume(int volume);
    public static native void screenshot(Bitmap bitmap);
    public static native GameInfo gameInfo(String path);

    private boolean inited;
    private boolean paused;
    public InputHandler handler;

    public YabauseRunnable(Yabause yabause)
    {
        handler = new InputHandler(this);
        int ok = init(yabause);
        Log.v("Yabause", "init = " + ok);
        inited = (ok == 0);
    }

    public void pause()
    {
        Log.v("Yabause", "pause... should really pause emulation now...");
        paused = true;
    }

    public void resume()
    {
        Log.v("Yabause", "resuming emulation...");
        paused = false;
        handler.post(this);
    }

    public void destroy()
    {
        Log.v("Yabause", "destroying yabause...");
        inited = false;
        deinit();
    }

    public void run()
    {
        if (inited && (! paused))
        {
            exec();

            handler.post(this);
        }
    }

    public boolean paused()
    {
        return paused;
    }
}

class YabauseHandler extends Handler {
    private Yabause yabause;

    public YabauseHandler(Yabause yabause) {
        this.yabause = yabause;
    }

    public void handleMessage(Message msg) {
        yabause.showDialog(msg.what, msg.getData());
    }
}

public class Yabause extends Activity implements OnPadListener
{
    private static final String TAG = "Yabause";
    private YabauseRunnable yabauseThread;
    private YabauseHandler handler;
    private YabauseAudio audio;
    private String biospath;
    private String gamepath;
    private int carttype;
    private PadManager padm;

    /** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        hideSystemUI();
        setContentView(R.layout.main);

        audio = new YabauseAudio(this);

        PreferenceManager.setDefaultValues(this, R.xml.preferences, false);
        readPreferences();

        Intent intent = getIntent();
        String game = intent.getStringExtra("org.yabause.android.FileName");

        if (game.length() > 0) {
            YabauseStorage storage = YabauseStorage.getStorage();
            gamepath = storage.getGamePath(game);
        } else
            gamepath = "";

        handler = new YabauseHandler(this);
        yabauseThread = new YabauseRunnable(this);

        padm = PadManager.getPadManager();

        YabausePad pad = (YabausePad) findViewById(R.id.yabause_pad);
        pad.setOnPadListener(this);
        if (padm.hasPad()) {
            pad.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    public void onPause()
    {
        super.onPause();
        Log.v(TAG, "pause... should pause emulation...");
        yabauseThread.pause();
        audio.mute(audio.SYSTEM);
    }

    @Override
    public void onResume()
    {
        super.onResume();
        Log.v(TAG, "resume... should resume emulation...");

        readPreferences();
        audio.unmute(audio.SYSTEM);

        yabauseThread.resume();
    }

    @Override
    public void onDestroy()
    {
        super.onDestroy();
        Log.v(TAG, "this is the end...");
        yabauseThread.destroy();
    }

    @Override
    public Dialog onCreateDialog(int id, Bundle args) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage(args.getString("message"))
            .setCancelable(false)
            .setNegativeButton("Exit", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                    Yabause.this.finish();
                }
            })
            .setPositiveButton("Ignore", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                    dialog.cancel();
                }
            });
        AlertDialog alert = builder.create();
        return alert;
    }

    @Override public boolean onPad(PadEvent event) {
        Message message = handler.obtainMessage();
        message.arg1 = event.getAction();
        message.arg2 = event.getKey();
        yabauseThread.handler.sendMessage(message);

        return true;
    }

    @Override public boolean onKeyDown(int keyCode, KeyEvent event) {
        PadEvent pe = padm.onKeyDown(keyCode, event);

        if (pe != null) {
            this.onPad(pe);
            return true;
        }

        return super.onKeyDown(keyCode, event);
    }

    @Override public boolean onKeyUp(int keyCode, KeyEvent event) {
        PadEvent pe = padm.onKeyUp(keyCode, event);

        if (pe != null) {
            this.onPad(pe);
            return true;
        }

        return super.onKeyUp(keyCode, event);
    }

    private void hideSystemUI() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
            getWindow().getDecorView().setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                | View.SYSTEM_UI_FLAG_FULLSCREEN
                | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
            );
        }
    }

    @Override public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);

        if (hasFocus) {
            hideSystemUI();
        }
    }

    private void errorMsg(String msg) {
        Message message = handler.obtainMessage();
        Bundle bundle = new Bundle();
        bundle.putString("message", msg);
        message.setData(bundle);
        handler.sendMessage(message);
    }

    private void readPreferences() {
        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(this);
        boolean fps = sharedPref.getBoolean("pref_fps", false);
        yabauseThread.enableFPS(fps ? 1 : 0);

        boolean frameskip = sharedPref.getBoolean("pref_frameskip", false);
        yabauseThread.enableFrameskip(frameskip ? 1 : 0);

        boolean audioout = sharedPref.getBoolean("pref_audio", true);
        if (audioout) {
            audio.unmute(audio.USER);
        } else {
            audio.mute(audio.USER);
        }

        String bios = sharedPref.getString("pref_bios", "");
        if (bios.length() > 0) {
            YabauseStorage storage = YabauseStorage.getStorage();
            biospath = storage.getBiosPath(bios);
        } else
            biospath = "";

        String cart = sharedPref.getString("pref_cart", "");
        if (cart.length() > 0) {
            Integer i = new Integer(cart);
            carttype = i.intValue();
        } else
            carttype = -1;
    }

    public String getBiosPath() {
        return biospath;
    }

    public String getGamePath() {
        return gamepath;
    }

    public String getMemoryPath() {
        return YabauseStorage.getStorage().getMemoryPath("memory.ram");
    }

    public int getCartridgeType() {
        return carttype;
    }

    public String getCartridgePath() {
        return YabauseStorage.getStorage().getCartridgePath(Cartridge.getDefaultFilename(carttype));
    }
}
